Allow custom filters in TissueMask

Christopher22 opened the pull request:
## Description

This small patch allows a custom "pipeline" of image filters within the TissueMask to optimize for specific demands regarding tissue. The default behavior is not changed.

## Types of Changes

<!--- What types of changes does your pull request introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Fixes: #375 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/histolab/histolab/blob/master/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

ernestoarbitrio commented on tests/unit/test_masks.py:
```diff
@@ -153,6 +154,17 @@ def it_knows_its_mask_tile(self, request):
 
         assert binary_mask == expected_mask
 
+    def it_knows_custom_image_filters(self):
+        tile = Tile(PILIMG.RGBA_COLOR_500X500_155_249_240, None, None)
+
+        # Create a tissue mask which should not cover anything
+        tissue_mask = TissueMask(
+            image_filters.RgbToGrayscale(), image_filters.OtsuThreshold()
+        )
+
+        binary_mask = tissue_mask(tile)
+        assert not np.any(binary_mask)
```
This test smells to me. Here you are not testing the functionality. A unit test need to mock the objects involved in the actual implementation, here you are asserting something that does not imply nothing.

Christopher22 replied:
Thank you for the review! To be honest, I could not completely get your point... With the unit test above, I just check that the result after applying the custom pipeline differs from the default (I might do that more explicit, as I rely on the current documentation there...) Testing the filters themselves is already covered in the other tests, is it not?

ernestoarbitrio replied:
My point is that a unit tests that has to proof that your changes are working. So maybe I should've been more clear, sorry. Given that I'd like to see a unit tests that mocks your objects you introduced. For example how do we know that `_CustomFiltersComposition` has been called properly? Ho can we know that the `setter` within that is doing the right thing? Is it working only on a `Slide` or on a `Tile` too?

My hunch is that the changes of the custom filters works only if applied on a `Slide` and not on a `Tile` I'd like to see a tests with this kind of proof.

IMHO i do not think that the test `it_knows_custom_image_filters` is not applying the filters you believe. So to disprove my hypothesis I'd like to see a tests with this kind of assertions.

Hope I've been clear 🤷🏽‍♂️ . Let me know 😄

Christopher22 replied:
You are of cause entirely right! I will go through the code again. The missing implementation for Tile is almost embarrassing ... 

Is there any documentation regarding your preferred unit tests and mock functions? They are amazingly sophisticated

ernestoarbitrio replied:
👍🏽 thanks a lot. Anyway there's no official preference doc for unit testing. I'd suggest a quick reading of https://realpython.com/python-mock-library/. Furthermore if you want you can join the slack channel (link in the README) and we can discuss there... or even we can have a brief call and we can do the implementation together. Let me know what do you think

Christopher22 replied:
Thank you very much for the additional information! I will join the Slack channel which seems like a valuable resource. 

I used my spare time on the train and added both the missing implementation and (hopefully) proper unit tests. 

What do you think?

ernestoarbitrio reviewed (changes requested):
Thanks for your contribution @Christopher22 ❤️ !
I have some concerns on this PR 🤷🏽‍♂️ . The first thing is the tests you provided. Please read my comment. 
Furthermore @alessiamarcolini can give you other consideration on you work.

ernestoarbitrio commented on histolab/filters/compositions.py:
```diff
@@ -139,3 +143,23 @@ def tissue_mask_filters(self) -> imf.Compose:
+        super().__init__()
+        self.custom_filters = custom_filters
+
+    def set_custom_filters(self, *custom_filters):
```
Please could you add the missing return type and the type for the `custom_filters`

ernestoarbitrio commented on histolab/filters/compositions.py:
```diff
@@ -139,3 +143,23 @@ def tissue_mask_filters(self) -> imf.Compose:
+
+
+class _CustomFiltersComposition(FiltersComposition):
+    def __init__(self, *custom_filters) -> None:
```
Please could you add the type for `custom_filters`?

ernestoarbitrio commented on histolab/filters/compositions.py:
```diff
@@ -41,15 +41,19 @@ class FiltersComposition:
     """
 
-    def __new__(cls: type, cls_):
+    def __new__(cls: type, cls_, *custom_filters):
```
Could you provide the missing typing annotation here?

ernestoarbitrio commented on histolab/filters/compositions.py:
```diff
@@ -139,3 +143,23 @@ def tissue_mask_filters(self) -> imf.Compose:
+    @lazyproperty
+    def tissue_mask_filters(self) -> imf.Compose:
+        """Filters composition for slide's tissue estimation with custom filters
+        Returns
```
```suggestion

        Returns
```

ernestoarbitrio commented on histolab/filters/compositions.py:
```diff
@@ -139,3 +143,23 @@ def tissue_mask_filters(self) -> imf.Compose:
+
+    @lazyproperty
+    def tissue_mask_filters(self) -> imf.Compose:
+        """Filters composition for slide's tissue estimation with custom filters
```
```suggestion
        """Filters composition for tissue estimation with custom filters.
```

ernestoarbitrio commented on histolab/masks.py:
```diff
@@ -134,8 +135,17 @@ def _regions(regions: List[Region], n: int = 1) -> List[Region]:
 
-    The tissue within the slide or tile is automatically detected through a predefined
-    chain of filters."""
+    By default, the tissue within the slide or tile is automatically
+    detected through a predefined chain of filters.
+    However, custom filters could be used, too.
```
I would move this part of the docstring from the class and place it within the `__init__`. Furthermore I'd get rid of the last line `However ... `

ernestoarbitrio commented on histolab/masks.py:
```diff
@@ -22,6 +22,7 @@
 import numpy as np
 
 import histolab
+from histolab.filters import image_filters
```
```suggestion
from .filters.image_filters import ImageFilter, MorphologicalFilter
```

ernestoarbitrio commented on histolab/masks.py:
```diff
@@ -134,8 +135,17 @@ def _regions(regions: List[Region], n: int = 1) -> List[Region]:
+    However, custom filters could be used, too.
+    """
+
+    def __init__(self, *filters: image_filters.ImageFilter):
```
```suggestion
    def __init__(self, *filters: List[Union[ImageFilter, MorphologicalFilter]]) -> None:
```

ernestoarbitrio commented on histolab/masks.py:
```diff
@@ -134,8 +135,17 @@ def _regions(regions: List[Region], n: int = 1) -> List[Region]:
+
+    def __init__(self, *filters: image_filters.ImageFilter):
+        """
+        Create a new tissue mask. If custom image filters are specified,
+        those are used instead of the default ones.
```
IMHO considering the https://www.python.org/dev/peps/pep-0257/ the first line of the docstring should be a sort of summary and fit in 88 chars. All the other info can be underneath the summary line with a black line between them.

Moreover the `Parameters` section of the docstring is missed 😄

ernestoarbitrio commented on histolab/masks.py:
```diff
@@ -178,7 +188,15 @@ def _mask(self, slide) -> np.ndarray:
         thumb = slide.thumbnail
-        filters = FiltersComposition(histolab.slide.Slide).tissue_mask_filters
+
+        # Generate appropiate composition of filter
```
```suggestion
        # Generate appropriate composition of filter
```

ernestoarbitrio commented on histolab/masks.py:
```diff
@@ -178,7 +188,15 @@ def _mask(self, slide) -> np.ndarray:
-        filters = FiltersComposition(histolab.slide.Slide).tissue_mask_filters
+
+        # Generate appropiate composition of filter
+        filters = (
+            FiltersComposition(histolab.slide.Slide)
+            if len(self.custom_filters) == 0
+            else FiltersComposition(
+                histolab.filters.image_filters.Compose, *self.custom_filters
+            )
+        ).tissue_mask_filters
```
Sorry but I see some problems in this portion of code:
1. `histolab.filters.image_filters.Compose` 👉🏽 could you import `Compose` without the whole reference of the histolab path?
2. The statement is too much convoluted maybe you should split it in something more readable (remember that the zen of python says: `beautiful is better than ugly 😂 `

ernestoarbitrio commented on histolab/masks.py:
```diff
@@ -197,4 +215,13 @@ def _(self, tile: Tile) -> np.ndarray:
+        # Otherwise, fall back to the default one.
+        if len(self.custom_filters) > 0:
+            custom_filters = FiltersComposition(
+                histolab.filters.image_filters.Compose, *self.custom_filters
```
As before 😄 ... import `Compose` and avoid the usage of the whole path.

ernestoarbitrio commented on histolab/tile.py:
```diff
@@ -168,7 +168,22 @@ def save(self, path: Union[str, bytes, os.PathLike]) -> None:
     @lazyproperty
     def tissue_mask(self) -> np.ndarray:
-        """Binary mask representing the tissue in the tile.
+        """Binary mask representing the tissue in the tile calculated
```
please keep the first line and the body separated with a blank line, and, as before, please follow the PEP 257 rules 🙏🏽 . Pay attention : The summary line must fit in one line (88chars)

ernestoarbitrio commented on histolab/tile.py:
```diff
@@ -168,7 +168,22 @@ def save(self, path: Union[str, bytes, os.PathLike]) -> None:
+    def calculate_tissue_mask(
+        self, filter_composition: FiltersComposition
+    ) -> np.ndarray:
+        """Calculate the binary mask representing the tissue in the tile
+        given FiltersComposition.
```
same as before for the PEP.

ernestoarbitrio replied:
```suggestion
        """Calculate the binary mask representing the tissue in the tile
        given ``filter_composition``.
```

ernestoarbitrio commented on tests/unit/test_compositions.py:
```diff
@@ -62,6 +66,24 @@ def it_knows_tissue_areas_mask_slide_filters_composition(
     assert type(_enough_tissue_mask_filters_) == Compose
 
 
+def it_knows_tissue_areas_mask_compose_filters_composition(
+    RgbToGrayscale_,
+    OtsuThreshold_,
+    BinaryDilation_,
+    RemoveSmallHoles_,
+    RemoveSmallObjects_,
+):
+    _enough_tissue_mask_filters_ = FiltersComposition(
+        Compose,
+        RgbToGrayscale_,
+        OtsuThreshold_,
+        BinaryDilation_,
+        RemoveSmallHoles_,
+        RemoveSmallObjects_,
+    ).tissue_mask_filters
+    assert type(_enough_tissue_mask_filters_) == Compose
```
IMHO this test is not meaningful. If you take a look at lines 52-58 you will see the assertions for the mocks. Could you please integrate those also in this test? Thanks

ernestoarbitrio commented on tests/unit/test_masks.py:
```diff
@@ -1,7 +1,11 @@
 import pytest
 
-from histolab.filters.compositions import _SlideFiltersComposition
+from histolab.filters import image_filters
```
please could avoid importing the whole module? I'd like to see the import statements only for the object needed 😄

ernestoarbitrio commented on tests/unit/test_masks.py:
```diff
@@ -138,6 +142,24 @@ def it_knows_its_mask_slide(
+    def it_knows_its_mask_slide_custom(self, request, tmpdir):
+        slide, _ = base_test_slide(tmpdir, PILIMG.RGBA_COLOR_500X500_155_249_240)
+        custom_filters = [image_filters.RgbToGrayscale()]
+
```
```suggestion
```

ernestoarbitrio commented on tests/unit/test_masks.py:
```diff
@@ -153,6 +175,39 @@ def it_knows_its_mask_tile(self, request):
 
+    def it_knows_its_mask_tile_and_compute_with_default(self, request):
+        tile = Tile(PILIMG.RGBA_COLOR_500X500_155_249_240, None, None)
+
```
```suggestion
```

ernestoarbitrio commented on tests/unit/test_masks.py:
```diff
@@ -153,6 +175,39 @@ def it_knows_its_mask_tile(self, request):
+    def it_knows_its_mask_tile_and_supports_custom_filters(self, request):
+        tile = Tile(PILIMG.RGBA_COLOR_500X500_155_249_240, None, None)
+        custom_filter_cls = image_filters.RgbToGrayscale
+
```
```suggestion
```

ernestoarbitrio commented on tests/unit/test_masks.py:
```diff
@@ -153,6 +175,39 @@ def it_knows_its_mask_tile(self, request):
+
+        custom_filters = [custom_filter_cls()]
+        custom_filter_call_ = method_mock(request, custom_filter_cls, "__call__")
+
```
```suggestion
```

ernestoarbitrio commented on tests/unit/test_masks.py:
```diff
@@ -138,6 +142,24 @@ def it_knows_its_mask_slide(
+        tissue_mask = TissueMask(*custom_filters)
+        _ = tissue_mask(slide)
+
+        set_custom_filters_.assert_called_once()
```
I'm not a huge fan of `assert_called_once`. The reason is that, yes, you're asserting that a proper method has been called, but the interesting info would be also know with which parameters? I'd suggest to switch to `assert_called_once_with` and check the calling args 😉

Christopher22 replied:
I try to remove all of the things. However, I sometimes have no access to the "self" parameter. In this case, I try to check only the remaining arguments.

ernestoarbitrio replied:
Sorry I do not understand this point. Could you elaborate? Thx

ernestoarbitrio commented on tests/unit/test_masks.py:
```diff
@@ -138,6 +142,24 @@ def it_knows_its_mask_slide(
+
+        set_custom_filters_.assert_called_once()
+        assert set_custom_filters_.call_args[0][1] == custom_filters[0]
+        custom_filter_call_.assert_called_once()
```
same

ernestoarbitrio commented on tests/unit/test_masks.py:
```diff
@@ -153,6 +175,39 @@ def it_knows_its_mask_tile(self, request):
+        tissue_mask = TissueMask()
+        _ = tissue_mask(tile)
+
+        calculate_tissue_mask_tile_.assert_called_once()
```
same

ernestoarbitrio commented on tests/unit/test_masks.py:
```diff
@@ -153,6 +175,39 @@ def it_knows_its_mask_tile(self, request):
+
+        # Ensure the custom tissue calculator, the internal setter
+        # and the custom filter itself is called
+        set_custom_filters_.assert_called_once()
```
same

ernestoarbitrio commented on tests/unit/test_masks.py:
```diff
@@ -153,6 +175,39 @@ def it_knows_its_mask_tile(self, request):
+        # and the custom filter itself is called
+        set_custom_filters_.assert_called_once()
+        assert set_custom_filters_.call_args[0][1] == custom_filters[0]
+        custom_filter_call_.assert_called_once()
```
same

alessiamarcolini replied:
Regarding the whole test: I feel that testing that `RgbToGrayscale` is called is not a good strategy, because:

1. usually we want to mock all the methods/instances that get called/used directly by the method we are testing, so in this case it would be `Tile.calculate_tissue_mask`
2. as a proof for my first point: this test is not really checking that the mask is built with the custom filters, because `RgbToGrayscale` is actually the first filter in the "default" filter sequence (for both `Slide` and Tile). So this test would actually pass with both the default and the "custom" filters represented by `[RgbToGrayscale()]` 🤓

Please let me know if this makes sense for you!

alessiamarcolini replied:
@Christopher22 I think you misunderstood my points here. This updated test covers point number 2 but not number 1, the mocked object is not directly called/used by the method you're testing.

ernestoarbitrio reviewed (changes requested):
@Christopher22 I did this CR round with @alessiamarcolini.
Sorry for the bunch of request changes ... but overall looks neat.
Could you please address the comments in order to make us able to do a final round?
Thanks again for you effort 💪🏽 ... we really appreciated it 😄

Christopher22 commented:
Thank you so much for your comments! I hope, I was able to fix the things up you mentioned.

As a small refactor, I removed my 'set_custom_filters' and replaced it with a meaningful constructor which fits better in your existing codebase.

ernestoarbitrio commented on histolab/masks.py:
```diff
@@ -178,8 +191,14 @@ def _mask(self, slide) -> np.ndarray:
+
+        # Generate appropriate composition of filter
+        if len(self.custom_filters) == 0:
+            composition = FiltersComposition(histolab.slide.Slide)
```
Could you import directly `Slide` and use it instead of the whole path?

ernestoarbitrio commented on tests/unit/test_masks.py:
```diff
@@ -148,11 +159,39 @@ def it_knows_its_mask_tile(self, request):
 
+    def it_knows_its_mask_tile_and_compute_with_default(self, request):
+        tile = Tile(PILIMG.RGBA_COLOR_500X500_155_249_240, None, None)
+
```
i would remove this line. The reason is that in a test, for the sake of readability and for future developers we're trying to keep separated the 3 blocks that compose a test.
1. Objects declarations
2. Actual call
3. Assertions

Hope this can help

ernestoarbitrio reviewed (changes requested):
Hey @Christopher22 thanks for all your commitment. I left a couple of furhter minor comments. Could you please take a look? After that we are probably ready to go.

Heads up to @alessiamarcolini ☝🏽

Christopher22 commented:
Amazing! Your highlighed areas should be fixed by now :)

alessiamarcolini commented on histolab/filters/compositions.py:
```diff
@@ -29,7 +31,7 @@ class FiltersComposition:
     Arguments
     ---------
-    cls_ : type, {Tile, Slide}
+    cls_ : type, {Tile, Slide, Compose}
```
docs: `*custom_filters parameter` is missing

alessiamarcolini commented on histolab/filters/compositions.py:
```diff
@@ -41,12 +43,13 @@ class FiltersComposition:
     """
 
-    def __new__(cls: type, cls_):
+    def __new__(cls: type, cls_, *custom_filters: imf.ImageFilter):
```
```suggestion
    def __new__(cls, cls_,: type *custom_filters: imf.ImageFilter):
```

alessiamarcolini commented on histolab/filters/compositions.py:
```diff
@@ -139,3 +142,35 @@ def tissue_mask_filters(self) -> imf.Compose:
+        """
+
+        if cls_ != imf.Compose:
+            raise FilterCompositionError("cls_ parameter cannot be None")
```
as Codecov says, this line is not covered by tests (in this case it should be a unit test)

alessiamarcolini commented on histolab/filters/compositions.py:
```diff
@@ -139,3 +142,35 @@ def tissue_mask_filters(self) -> imf.Compose:
+            raise FilterCompositionError("cls_ parameter cannot be None")
+
+        if len(custom_filters) == 0:
+            raise FilterCompositionError("'Compose' requires at least one filter")
```
same here, not covered by tests

alessiamarcolini commented on histolab/filters/compositions.py:
```diff
@@ -139,3 +142,35 @@ def tissue_mask_filters(self) -> imf.Compose:
+        imf.Compose
+            Filters composition
+        """
+        if len(self.custom_filters) == 0:
```
Is this possible? If yes, it needs a test

alessiamarcolini commented on histolab/masks.py:
```diff
@@ -132,10 +133,22 @@ def _regions(regions: List[Region], n: int = 1) -> List[Region]:
+
+        Parameters
+        ----------
+        filters : ImageFilter
```
```suggestion
        *filters : ImageFilter
```
Otherwise I wouldn't know how to specify that that's an arbitrary number of parameters

alessiamarcolini commented on tests/unit/test_masks.py:
```diff
@@ -148,11 +159,38 @@ def it_knows_its_mask_tile(self, request):
         tissue_mask_tile_.return_value = expected_mask
 
```
not your responsibility, but since you touched this method... this line should also be removed

alessiamarcolini commented on tests/unit/test_masks.py:
```diff
@@ -148,11 +159,38 @@ def it_knows_its_mask_tile(self, request):
+        tissue_mask = TissueMask()
+        _ = tissue_mask(tile)
+
+        calculate_tissue_mask_tile_.assert_called_once()
```
the problem of not checking the parameters can hide bugs in this very case: this test is not really checking that the mask is computed with the default filters. The method `Tile.calculate_tissue_mask` could be called:
1. by `Tile.tissue_mask`, with `FiltersComposition(Tile)` as parameter, if the code logic is correct OR
2. by _mask (dispatched to the `Tile`'s method `_`) with `custom_filters`, if the logic is not correct.

So please add the parameters in the assertion :)

alessiamarcolini reviewed (changes requested):
Apart from a couple of docstrings improvements, I had to reason carefully about two unit tests (see comments) and please try to cover all of your new code with tests (follow codecov comments), as we cannot merge this PR if it does not provide 100% coverage ✌🏼

<img src="https://media2.giphy.com/media/wA6aYAO3jH6RmUIBYs/giphy.gif"/>

Christopher22 commented:
Such cute apes! ;) Your comments should be integrated by now!

alessiamarcolini reviewed (changes requested):
Please check the comment I left about the `it_knows_its_mask_tile_and_supports_custom_filters` test.

PS please squash the commits at the end of the review, otherwise it's more difficult for us to be on track with the changes between review rounds ✌🏼

Christopher22 commented:
@alessiamarcolini Do you prefer a test such as this instead? :)

alessiamarcolini commented on tests/unit/test_masks.py:
```diff
@@ -177,22 +180,30 @@ def it_knows_its_mask_tile(self, request):
-        # We need to account for the fact padding is added before filtering
-        expected_mask = np.zeros((510, 510), dtype=np.bool)
-        custom_filter_call_.return_value = expected_mask
+        expected_mask = [
+            [True, True],
+            [False, True],
+        ]
```
please use a np array here to be consistent with the real `calculate_tissue_mask` return type

alessiamarcolini commented on tests/unit/test_masks.py:
```diff
@@ -177,22 +180,30 @@ def it_knows_its_mask_tile(self, request):
+            custom_filters
+        )
+
+        assert binary_mask == expected_mask
```
accordigly, here please use `np.testing.assert_array_equal`

alessiamarcolini reviewed (changes requested):
Just a minor comment for consistency, the rest is perfect!! 👏🏼

<img src="https://media3.giphy.com/media/8UF0EXzsc0Ckg/giphy.gif"/>

Christopher22 commented:
Ready for squash ;)

alessiamarcolini reviewed (approved):
Let's merge it!

<img src="https://media1.giphy.com/media/fA81FF4mdE6lgeoJwb/giphy.gif"/>

alessiamarcolini commented:
@Christopher22 thank you so much for your valuable contribution! 👏🏼👏🏼

Now, to have a consistent code base, we would like to provide the same flexibility for the `BiggestTissueBoxMask` as well. Please let us know if you are interested in contributing again in this regard! Otherwise, we can take care of it, no pressure at all 😉
